content/java-interview-questions-2/4-new-java-features.md

```diff
@@ -1,37 +1,96 @@
+ - We can use @FunctionalInterface annotation for these interfaces.
+---
 
-# What is the interviewer looking for?
-- One
-- Two
-- Three
+# Why the static or default methods in interface are introduced in Java 8 ?
+--
+# Answer
+ -  In the java old version there is a lot of use of interfaces, and to provide the backword compatibility for these interfaces, default methods are introduced. 
+ - Static methods are also same only thing you can not override that method so it kind of provide new feature 
+---
 
+# Can you explain few stream operations ?
 --
+# Answer
+ - Stream operations are typically divided into 2 parts
+ - Intermediate operations returns only Stream object, while terminal operations returns some definite type of object.
+   1) Intermediate operations     - filter(), map(), flatmap()
+   2) Terminal operations         - anyMatch(), allMatch(), collect()
+ 
+---
 
+# Can you differentiate between sequential stream and parallel stream ?
+--
 # Answer
-- One
-- Two
-- Three
+  - primary difference between these streams is that sequential stream processes the data in sequencial order while parallel stream processes data in random order. 
+  - Parallel stream makes use of processor cores for processing the data, so parallel stream will work effectively when your machine has multiple cores.
+  - Also you should use parallel stream when have very large data to process and tasks are independant.
+---
 
+# What is Optional and why do we need it ?
+--
+# Answer
+  - Optional is new class introduced in Java8, it provides way to handle the nulls in program. 
+  - As we are working on objects there is a chance that we could have null objects in our program, which causes NullPointerException as effect it will terminate the program. 
+    This is big issue if it's production environment, to prevent these kind of incident Optional is introduced.
 ---
 
-## Question
-# What is Spring?
+# What is the change in Date time APIs ?
+--
+# Answer
+  - Old API was not a thread safe and has less operations. 
+  - In Java 8 date time API we have multiple operations and flexibility, it also introduced the zoned date time.
+---
 
+# What is Lambda expression and why it's introduced ?
 --
+# Answer
+  - Lambda expression is nothing but a piece of code or you can a small logic which can be provided as parameter to a method. 
+  - Syntax of lambda is as below
+    ``` 
+    () -> { statement;} 
+    ```
+  - Lambda expression is very powerfull feature, it allows use to execute code/logic without creating the object of the class. 
+  - Also to enable the use of functional interfaces we need lambda expression. This is way of functional programming and java is heading towards functional programming.
+---
 
-# What is the interviewer looking for?
-- One
-- Two
-- Three
+# how do you break the forEach loop of stream ?
+--
+# Answer
 
+---
+# What is Predicate ?
 --
+# Answer
 
+---
+
+# What is Consumer function ?
+--
 # Answer
-- One
-- Two
-- Three
 
```

Comment:
Please remove any questions that don't have answers. We can add them when they are ready.